libs/x-content/src/main/java/org/elasticsearch/xcontent/Text.java

```diff
@@ -36,31 +31,45 @@ public static Text[] convertFromStringArray(String[] strings) {
         return texts;
     }
 
-    private BytesReference bytes;
+    private EncodedBytes bytes;
     private String text;
     private int hash;
+    private int stringLength = -1;
+
+    /**
+     * Construct a Text from a UTF-8 encoded ByteBuffer. Since no string length is specified, {@link #stringLength()}
+     * will perform a string conversion to measure the string length.
+     */
+    public Text(EncodedBytes bytes) {
+        this.bytes = bytes;
+    }
 
-    public Text(BytesReference bytes) {
+    /**
+     * Construct a Text from a UTF-8 encoded ByteBuffer and an explicit string length. Used to avoid string conversion
+     * in {@link #stringLength()}.
```

Comment:
I think the javadocs here should answer this question: Is there a requirement that the `stringLength` matches the value that would have been produced by the other constructor? Or is the caller free to specify a different length?